tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc

```diff
@@ -1453,39 +1453,181 @@ Status Converter::RenameAndMarkOutputTensors(
+static constexpr int32_t kLAYER_IMPL_BASE = 0x80000000;
+#else
+static constexpr int32_t kLAYER_IMPL_BASE = 0x00000000;
+#endif
+
+enum class LayerImpl : int64_t {
+#if IS_TRT_VERSION_GE(8, 0, 0, 0)
+  kSHUFFLE = kLAYER_IMPL_BASE + 14
+#else
+  kSHUFFLE = kLAYER_IMPL_BASE + 16
+#endif
+};
 
+// An algorithm selector to support debugging and work around known TensorRT
+// issues.
+class TftrtAlgorithmSelector : public nvinfer1::IAlgorithmSelector {
+  // A list of tactics allowed to be selected. An empty list means that all
+  // tactics are allowed. This list is initialized from environment variable
+  // TF_TRT_FIXED_ALGORITHM_ID.
+  std::unordered_set<int64> allowed_algorithm_ids;
+
+  // A list of tactics that we disallow. Used primarily to go around known
+  // TensorRT issues.
+  std::unordered_set<int64> disallowed_tactics;
  public:
-  StaticAlgorithmSelector(int32_t algorithm_id) : algorithm_id_(algorithm_id) {}
+  TftrtAlgorithmSelector() {
+    int64_t trt_algorithm_id;
+    int64_t null_tactic = std::numeric_limits<int64_t>::min();
+    TF_CHECK_OK(tensorflow::ReadInt64FromEnvVar("TF_TRT_FIXED_ALGORITHM_ID",
+                                                /*default_val=*/null_tactic,
+                                                &trt_algorithm_id));
+    // Populate allowed tactics. Used to force select tactics for any layer.
+    // TODO: accepting a list from the environment variable
+    if (trt_algorithm_id != null_tactic) {
+      allowed_algorithm_ids.insert(trt_algorithm_id);
+    }
 
+#if !IS_TRT_VERSION_GE(8, 0, 0, 0)
+    disallowed_tactics = {
+      // turing_fp16_s1688cudnn_fp16_128x128_ldg8_relu_f2f_exp_medium_nhwc_gelu_tn_v1
+      -5927686925093575778,
+      //turing_fp16_s1688cudnn_fp16_128x128_ldg8_relu_f2f_exp_interior_nhwc_gelu_tn_v1
+      -3848538574386518527,
+      //turing_fp16_s1688cudnn_fp16_128x128_ldg8_relu_f2f_exp_small_nhwc_gelu_tn_v1
+      -959009792490796596
+    };
+#endif  // !IS_TRT_VERSION_GE(8, 0, 0, 0)
+
+    // Check whether allowed and disallowed tactics are disjointed.
+    if (allowed_algorithm_ids.size() > 0 && disallowed_tactics.size() > 0) {
+      std::vector<int64_t> intersection;
+      for (int64_t allowed : allowed_algorithm_ids) {
+        if (disallowed_tactics.count(allowed) > 0) {
+          intersection.emplace_back(allowed);
+        }
+      }
+      if (intersection.size() > 0) {
+        VLOG(1) << "Intersection of allowed and disallowed tactics found "
+                   "non-empty.";
+
+        VLOG(1) << "List of allowed tactics:";
+        VLOG(1) << absl::StrJoin(allowed_algorithm_ids, ", ");
+
+        VLOG(1) << "List of disallowed tactics:";
+        VLOG(1) << absl::StrJoin(disallowed_tactics, ", ");
+
+        VLOG(1) << "List of tactics common to both allowed and disallowed "
+                   "tactics:";
+        VLOG(1) << absl::StrJoin(intersection, ", ");
+
+        VLOG(1) << "Aborting.";
+        assert(0);
+      }
+    }
+  }
```

Comment:
This code block seems to contradict with the code in lines 1571-1586.
Tactics and algorithm_ids are not talking about the same thing, they can have the same value but refer to different things. Right?

Reply:
Both algorithm_id and tactic represent the same thing. Do you prefer to replace references to tactic to algorithm_id in the code base to reduce confusion?